src/llfs/storage_file_builder.test.cpp

```diff
@@ -56,6 +56,114 @@ class StorageFileBuilderTest : public ::testing::Test
+            llfs::config_slot_cast<llfs::PackedPageDeviceConfig>(slot);
+        ASSERT_EQ(page_device_config.page_0_offset, expected_page0_offset);
+        ASSERT_EQ(page_device_config.device_id, expected_device_id);
+        ASSERT_EQ(page_device_config.page_count, (little_i64)kTestPageCount);
+        ASSERT_EQ(page_device_config.page_size_log2, page_size_log2);
+
+        // Update page offset numbers and device ID
+        //
+        expected_page0_offset +=
+            page_device_config.page_size() * kTestPageCount - llfs::PackedConfigSlot::kSize;
+        expected_device_id += 1;
+      }
+
+      // Assuming 1 slot filled if back slot, all slots filled otherwise.
+      //
+      constexpr u64 kMaxPageDeviceSlots =
+          llfs::PackedConfigBlock::kPayloadCapacity / llfs::PackedPageDeviceConfig::kSize;
+      u64 expected_num_slots = (config_block == config_blocks.back()) ? 1 : kMaxPageDeviceSlots;
+
+      // Set offsets of config block based on position
+      //
+      if (config_block == config_blocks.front()) {
+        expected_prev_offset = llfs::PackedConfigBlock::kNullFileOffset;
+      } else {
+        expected_prev_offset = -expected_next_offset;
+      }
+      if (config_block == config_blocks.back()) {
+        expected_next_offset = llfs::PackedConfigBlock::kNullFileOffset;
+      } else {
+        expected_next_offset = batt::round_up_bits(
+            12, expected_page0_offset + llfs::PackedPageDeviceConfig::kSize * expected_num_slots);
+      }
+
+      // Verify the packed config block
+      //
+      ASSERT_EQ(p_config_block.magic, llfs::PackedConfigBlock::kMagic);
+      ASSERT_EQ(p_config_block.version, llfs::make_version_u64(0, 1, 0));
+      ASSERT_EQ(p_config_block.prev_offset, expected_prev_offset);
+      ASSERT_EQ(p_config_block.next_offset, expected_next_offset);
+      ASSERT_EQ(p_config_block.slots.size(), expected_num_slots);
+      ASSERT_EQ(p_config_block.crc64, p_config_block.true_crc64());
+    }
+  }
+
+  // Writes random data to disk given the file, offset, and amount of data to write. Used to
+  // randomize data of a test file to verify repeated write and read testing.
+  //
+  void write_rand_data(llfs::IoRingRawBlockFile& test_file, i64 write_offset, i64 write_size)
+  {
+    constexpr i64 kBufferSize = 32768;
+    i64 write_bytes_remaining = batt::round_up_bits(9, write_size);
+    i64 offset = batt::round_up_bits(llfs::PackedConfigBlock::kSize, write_offset);
+
+    llfs::StatusOr<int> rand_fd = llfs::open_file_read_only("/dev/random");
+    ASSERT_TRUE(rand_fd.ok()) << BATT_INSPECT(rand_fd.status());
+
+    const auto closer = batt::finally([rand_fd] {
+      llfs::Status close_status = llfs::close_fd(*rand_fd);
+      ASSERT_TRUE(close_status.ok()) << BATT_INSPECT(close_status);
+    });
+
+    std::aligned_storage_t<kBufferSize, 512> rand_data;
+
+    while (write_bytes_remaining > 0) {
+      llfs::MutableBuffer read_buffer(&rand_data, kBufferSize);
```

Comment:
AFAICT, the code creates `read_buffer` only to just pull out the pointer passed in as the first ctor arg (next line, `read_buffer.data()`, so I am wondering why create this object at all?  When we spoke a while back about the advantages of using `MutableBuffer`/`ConstBuffer` over pairs of function arguments (pointer and size), the point is to pass a single arg to some function instead of 2.  But here, the code is using POSIX `read`, which doesn't do that.  `<llfs/filesystem.hpp>` does define `read_fd`, which is more like `pread` than `read` in that it includes an extra `offset` arg; would that work (with "/dev/random")?  If not, then might it make more sense to add a new function to filesystem.hpp, perhaps an overload of `read_fd` that drops the last parameter?  (`u64 offset`)

Reply:
No longer using <llfs/filesystem.hpp> to handle reading data. Using the <random> API to populate a read MutableBuffer which takes advantage of managing the pointer and size since it take several iterations to fill.

Dont feel there is any need to make an overload for llfs::read_fd.